### ISSUE TITLE
keyword arguments for x0 in stats.rv_continuous

### ISSUE THREAD

**Issue by vyorgan** [scipy/scipy#5362]
When fitting a distribution, it would be useful to be able to specify shape parameter starting values with keywords instead of args- eg

```
scipy.stats.gamma.fit(mydata, floc=0, a=2)
```

This is already allowed for scale and loc, but not for the shape params. It's easy to change, such as

```
if self.shapes:
     shapes = self.shapes.replace(',', ' ').split()
     for i, s in enumerate(shapes):
           args[i]=kwds.pop(s, args[i])
```

in either the rv_continuous.fit or rv_continuous._reduce_func

**Comment by ev-br** [comment]
Do you want to fix a (floc fixes the loc etc) and vary the scale only, or fix the location and fit the scale and a using 2.0 as a starting value for a?

In the former case you can use fa=2 ( equivalently, fix_a=2).

**Comment by vyorgan** [comment]
Thanks. No, I don't want to fix variables in this case, I just want to set their initial values for the optimizer. I know now I can use args, but it is more convenient to use kwargs. In a distribution with one shape parameter it's trivial, but if you have more, it's less error-prone to go by meaningful labels than the arg order (or I can use **params with the fixed and free variables in a dict).

I made the error of trying to initialize them with keywords (like you can with scale and loc)- it led to frustration because it was not finding global MLE, no matter what (I thought) my starting x0s were because it was just ignoring those inputs. I figured if I had this confusion, others might, as well.

**Comment by ev-br** [comment]
Yes, it would definitely be a very welcome enhancement. Are you interested in contributing a patch?

One thing to do here could be to improve input validation --- just like you're saying, it silently ignores arguments, and this not good. Maybe the easiest would be to cook up a parameter-parsing function at import time, similar to` _parse_args`. Or maybe not, I did not think about it much.

**Comment by WarrenWeckesser** [comment]
@vyorgan wrote:

> it was just ignoring those inputs

You should understand that a few of the distributions use explicit formulas to compute the MLE fit, and in those cases, the initial conditions are ignored, because they are not needed.  `gamma.fit` with a given `floc` is one of those cases.  There is an explicit formula for the scale, and the shape parameter is found by solving a single implicit equation.  (See my comment at https://github.com/scipy/scipy/pull/2519#issuecomment-22419895 for more details.)  So even if the `fit` method correctly handled `a=2`, it would still be ignored.  If you have a case where `gamma.fit` is returning an incorrect result, please file a bug report (including a runnable example that demonstrates the problem, if possible).

So there is a documentation problem: the docstrings of the `fit` method for those distributions that use explicit formulas should explain the cases in which the initial conditions are ignored.

By the way: in master, the problem of `gamma.fit` ignoring the keyword argument `a=2` has been "fixed"--at least now you'll get an error:

```
In [17]: scipy.__version__
Out[17]: '0.17.0.dev0+619d5b7'

In [18]: gamma.fit([1, 2, 3, 4, 3, 1.5, 0], a=1.0)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-18-19f0b807a9bb> in <module>()
----> 1 gamma.fit([1, 2, 3, 4, 3, 1.5, 0], a=1.0)

/Users/warren/local_scipy/lib/python2.7/site-packages/scipy/stats/_continuous_distns.pyc in fit(self, data, *args, **kwds)
   1908         if floc is None:
   1909             # loc is not fixed.  Use the default fit method.
-> 1910             return super(gamma_gen, self).fit(data, *args, **kwds)
   1911 
   1912         # Special case: loc is fixed.

/Users/warren/local_scipy/lib/python2.7/site-packages/scipy/stats/_distn_infrastructure.pyc in fit(self, data, *args, **kwds)
   2139         # by now kwds must be empty, since everybody took what they needed
   2140         if kwds:
-> 2141             raise TypeError("Unknown arguments: %s." % kwds)
   2142 
   2143         vals = optimizer(func, x0, args=(ravel(data),), disp=0)

TypeError: Unknown arguments: {'a': 1.0}.
```

Enhancing the `fit` method by allowing the initial conditions for the shape parameters to be specified using keywords sounds like a good idea, but they will still be ignored in some cases.  If this feature is added, perhaps the case where the values are ignored should generate a warning, or even an error.

**Comment by WarrenWeckesser** [comment]
*Correction!*  The problem is not "fixed" in master; my example just happened to work because of the specialized code in `gamma.fit`.  Here's what happens when you specify `floc=0, a=1` (with data that is strictly positive):

```
In [22]: gamma.fit([1, 2, 3, 4, 3, 1.5, 0.5], floc=0, a=1.0)
Out[22]: (2.7556210111399833, 0, 0.77763129769817518)
```
`a=1.0` is silently ignored.

**Comment by vyorgan** [comment]
Interesting- I used gamma as an arbitrary example, but it explains some unexplained behaviors I had before (ignoring incorrect keywords). I think it's a separate issue -whether the explicit MLE ignores x0 or not- than my confusion. I think (if I understand) what you are describing is already a problem with the current scheme if you specify 'a' in the args, eg
```
gamma.fit([1, 2, 3, 4, 3, 1.5, 0.5], 'ignored_a', floc=0)
```
runs fine. It also appears to ignore any unexpected keywords, such as 'optimizer.' 

I don't care too much if it ignores my x0 when it has an exact solution (when my inputs don't make a difference, anyway). I think if I specified an optimizer, I would want it use that instead of the exact method. I wouldn't want it to throw an error for keywords that would work under other circumstances (a different distribution, an unfixed loc, etc), because that would be inconsistent and confusing to me.

As to a contribution, I'm new to github and not exactly sure how to do that yet- for my purposes I overwrote 'fit' in my own subclass of 'rv_continuous' for the distributions I needed, which looks like this- 

```
class compoundDist(rv_continuous):
#other methods here...
    def fit(self, data, *args, **kwds):
        args=list(args)
        Narg = len(args)
        if Narg > self.numargs:
            raise TypeError("Too many input arguments.")

        start = [None]*2
        if (Narg < self.numargs) or not ('loc' in kwds and
                                         'scale' in kwds):
            # get distribution specific starting locations
            start = self._fitstart(data)
            args += start[Narg:-2]
        loc = kwds.pop('loc', start[-2])
        scale = kwds.pop('scale', start[-1])
        if self.shapes:
            shapes = self.shapes.replace(',', ' ').split()            
            for i, s in enumerate(shapes):
                args[i]=kwds.pop(s, args[i])
        return super(compoundDist, self).fit(data, *args[:-2], **kwds)
```        

The first part was copied out of fit, and the behavior should be the same if the just the last 4 lines were added into it at the corresponding location.

**Comment by ev-br** [comment]
> As to a contribution, I'm new to github and not exactly sure how to do that yet- 

Have a look at the NumPy guide, it is (mostly) applicable: http://docs.scipy.org/doc/numpy/dev/gitwash/index.html

Regarding specifics of the feature, I'd recommend starting with writing tests to fix the interface. Existing tests for the fit method are here, https://github.com/scipy/scipy/blob/master/scipy/stats/tests/test_distributions.py#L1099
Note that we need to keep backwards compatibility, so that what works now must keep working with your changes (unless there are serious reasons to change the behavior, and that'd be a contributor's job to argue for the change). 
Also note that the current test coverage is not exhaustive, so it might make sense to add some tests just to fix the current behavior.
Once the interface is fixed by tests (and a bunch of them fail), you start implementing the actual feature until all tests pass. When they do, you're done :-). 

Also note that you can send a work-in-progress (WIP) pull requests. If you get stuck somewhere, ping us.

**Comment by mdhaber** [comment]
@vyorgan would you consider [`scipy.stats.fit`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.fit.html) to meet this need? 

```python3
from scipy import stats
guess = {'a': 2}
bounds = dict(a=[0.1, 10], loc=[0, 0], scale=[0.1, 10])
data = stats.gamma(2.1, scale=0.5).rvs(100)
stats.fit(stats.gamma, data, bounds=bounds, guess=guess)

# FitParams(a=np.float64(1.8809796881634446), loc=np.float64(0.0), scale=np.float64(0.5112968374982736))
# success: True
# message: 'Optimization terminated successfully.'
```
